pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotRuleUtils.java

```diff
@@ -68,19 +69,28 @@ public static boolean isAggregate(RelNode rel) {
   }
 
   // TODO: optimize this part out as it is not efficient to scan the entire subtree for exchanges.
-  public static boolean noExchangeInSubtree(RelNode relNode) {
-    if (relNode instanceof HepRelVertex) {
-      relNode = ((HepRelVertex) relNode).getCurrentRel();
-    }
-    if (relNode instanceof Exchange) {
+  public static boolean canPushDynamicBroadcastToLeaf(RelNode relNode) {
```

Comment:
change this to positive-search: only allow nodes that doesn't change row-semantics 
- JOIN/PROJECT/FILTER/TABLE_SCAN
- technically WINDOW and SORT also doesn't change row-semantics but SORT could have limit and WINDOW is not implemented on leaf. will follow up later